Eliminate non-escaping boxes, tuples and refs

lrytz opened the pull request:
This PR includes the commits of #4846.

Eliminate boxes, tuples and refs that are created and used within a
single method without escaping. For details on the implementation see
the doc comment in class BoxUnbox.

lrytz committed:
Track aliases of primitive values in AliasingFrame

This allows using an AliasingAnalyzer for copy propagation
(subsequent commit).

lrytz committed:
[trivial] import Opcodes._ in two files, move a utility method

lrytz committed:
Fix VarInstruction extractor to include IINC

Previously the VarInstruction extractor did not include IINCs.
Also give rename the isVarInstruction predicate to isLoadStoreOrRet,
as it doesn't include IINC.

Also fixes a bug in removeJumpAndAdjustStack, it checked the wrong
opcode range before.

lrytz committed:
Remove InstructionResultSize (subsumed by InstructionStackEffect)

lrytz committed:
Clean up DCE: remove eliminated callsites from call graph earlier

When running DCE, directly remove eliminated callsites from the call
graph (instead delegating this task to the callees).

lrytz committed:
Fix merging of AliasingFrames

By convention (for performance), `aliasingFrame.aliases(slot) == null`
means that the value at `slot` has no aliases, so this is isomorphic
to having a singleton `AliasSet(slot)`.

When merging an AliasingFrame `other` into `this`, at every value slot
we have to keep the intersection of the two alias sets. In other words
we have to remove from `this.aliases(slot)` those values that are
not in `other.aliases(slot)`. This was wrong for the cases where
`other.aliases == null` - we did not treat this case as a singleton
set.

lrytz committed:
Copy propagation, remove unused values (closures!) and local variables

Copy propagation uses an AliasingAnalyzer: it replaces a `LOAD n`
instruction by `LOAD m` where m is the smallest alias of n. This
leads to stale STORE instructions.

Stale STOREs are identified using a ProdCons analyzer and replaced by
POPs.

Values that are pushed on the stack by a side-effect free instruction
and consumed by a POP are then removed by `eliminatePushPop`. This
includes elimination of unused closure allocations and unused boxes
and  tuple allocations (*).

A final cleanup eliminates `STORE x; LOADx` pairs where the stored
value is not otherwise used.

Fixes
  - https://github.com/scala/scala-dev/issues/25
  - https://github.com/scala/scala-dev/issues/7
  - https://github.com/scala/scala-dev/issues/14
  - https://github.com/scala/scala-dev/issues/12

(*) We don't yet rewrite reads of boxes and tuples yet. For example,
`val x = (1, 2); x._1` remains a method invocation and the tuple
cannot be eliminated (https://github.com/scala/scala-dev/issues/11).

Inspired in many ways by Miguel's work!

lrytz committed:
Clean up optimizer settings.

lrytz committed:
Fix specialized function class names in closure optimizer

lrytz committed:
Support specialized functions in closure optimizer

Fixes https://github.com/scala/scala-dev/issues/52.

An IndyLambda may create a specialized function type, where the SAM
is the corresponding specialized variant of apply. If this closure
is invoked through the generic apply method, the closure optimizer
would previously not re-write the invocation to the $anonfun method.
This is now done, including the necessary box / unbox operations.

lrytz committed:
Fix in "consumersOfOutputsFrom" query in prod-cons analysis

Fix the "consumersOfOutputsFrom" in the prod-cons analysis to support
dummy instructions:

  - UninitializedLocalProducer
  - ParameterProducer
  - ExceptionProducer

lrytz committed:
Rename test MethodLevelOpts -> MethodLevelOptsTest

lrytz committed:
Eliminate non-escaping boxes, tuples and refs

Eliminate boxes, tuples and refs that are created and used within a
single method without escaping. For details on the implementation see
the doc comment in class BoxUnbox.

This commit also cleans up the logic of inter-dependent method-level
optimizations that run until reaching a fixpoint.

lrytz committed:
Make some testing tools from InlinerTest available to other tests

lrytz committed:
Minor cleanups in eliminateStaleStores

lrytz committed:
Apply local optimization based on nullness information

Optimize IFNULL branches to GOTO when (non-)nullness of the tested
value is known statically. This enables unreachable code to be
removed, which in turn enables boxes to be eliminated.

Changed a test flag from `-Ynooptimise` to `-Yopt:l:classpath` - I
still have to do this systematically, this will follow later.

lrytz committed:
Fix push-pop elimination for values pushed by DUP

If a DUP is consumed by two POPs, ensure that the DUP and its producer
are eliminated. Before, only the DUP was eliminated, leaving an unused
value on the stack.

lrytz committed:
Don't run closure optimizer on methods too large for prod-cons

For methods that are too large to run a producers-consumers analysis,
don't run the closure optimizer.

This is an oversight, all data flow analyses in the backend are
guarded by a method size check.

lrytz committed:
Ensure that `-Yopt:l:none` takes precedence, tidy optimizer settings

lrytz committed:
Eliminate unnecessary casts

Eliminate casts that are statically known to succeed. This enables
boxes to be eliminated and simplifies the implementation of closure
allocation elimination.

lrytz committed:
Harden push-pop elimination when eliminating constants

Ensure that LDC instructions are only eliminated when the loaded value
is a numeric or string constant. Removing other literals may remove
a potential failure. [1] lists the kind of values that can be loaded
by a LDC

[1] https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.ldc

lrytz commented:
@retronym this is now ready for review. This PR includes the commits of https://github.com/scala/scala/pull/4837 and https://github.com/scala/scala/pull/4846, so you can review everything here if you like. I addressed the comments you gave in 4837. A community build is running here: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/269/parameters/

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -140,6 +145,14 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
 
+  def getBoxedUnit: FieldInsnNode = new FieldInsnNode(Opcodes.GETSTATIC, coreBTypes.srBoxedUnitRef.internalName, "UNIT", coreBTypes.srBoxedUnitRef.descriptor)
+
+  private val anonfunAdaptedName = """.*\$anonfun\$\d+\$adapted"""
```
Precompile the regex here?

```
scala> val r = "foo..".r
r: scala.util.matching.Regex = foo..

scala> r.pattern.matcher("fooab").matches
res9: Boolean = true
```

lrytz replied:
:+1:

lrytz commented:
community build status

  - Shapeless failed because it has `-Xfatal-warnings` enabled and there's an invocation to `List.dropWhile` that cannot be inlined, producing the `there was 1 inliner warning` summary. The method cannot be inlined because it contains an `INVOKESPECIAL` of a private method (the `loop` method is private, see https://github.com/scala/scala/blob/v2.12.0-M3/src/library/scala/collection/immutable/List.scala#L354-L361). The fix here is to pass `-Yopt-warnings:none`.
  - The shapeless failure looks a bit more dramatic than it is because the compiler crashes after issuing the error, due to https://issues.scala-lang.org/browse/SI-9592.
  - Scala.js failed, looking at it now.

lrytz committed:
Run DCE before the closure optimizer (fixes a crash)

Before identifying function callsites within the same method as a
closure allocation, run DCE. The ProdCons analysis used to identify
these function calls may crash if there is unreachable code, as
observed in the community build with scala-js.

The crash was rare because inlining, which is performed before closure
optimizations, already runs DCE. However, inlining may render more
code unreachable (e.g. when inlining a method that throws).

Also make sure that DCE is always performed on the callee before
inlining: move the DCE invocation into the inlineCallsite method,
which is also invoked by the closure optimizer.

lrytz committed:
Move copy propagation implementation to a separate file

lrytz commented:
/rebuild

lrytz commented:
Community build passed with `-Yopt:l:classpath` (https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/277/)

retronym commented:
Any chance for a unit test for "Run DCE before the closure optimizer (fixes a crash)" ? Getting pretty close to :santa: on this PR otherwise!

lrytz commented:
I'll give it a try, have to think about it. The bug was discovered in the community build on scala-js. I didn't isolate / minimize it, the stack trace and some mental processing was enough to realize what is the problem.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala:
```diff
@@ -311,13 +341,298 @@ object BytecodeUtils {
+    } ++ {
+      for (r <- ijfdzv; a <- ijd; b <- ijd) yield base + "2$mc" + r + a + b + "$sp" // Function2
+    }
+  }
```
The string refs to symbols in this file, and to the specialized ones here in particular, are the main thing I think needs a cleanup before merging.

lrytz replied:
ok, i'll look into cleaning this up

lrytz commented:
> Any chance for a unit test for "Run DCE before the closure optimizer (fixes a crash)" ?

pushed (amendeded last commit,  4eee5e0)

lrytz committed:
Precompile some regexes used in the backend

lrytz commented:
@retronym regarding the enumeration of specialized subclasses of `FunctionN` [(here)](https://github.com/scala/scala/blob/4eee5e088eb72ba00aa4b02a03a1bcf27b5d751b/src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala#L613-L628) / `TupleN` [(here)](https://github.com/scala/scala/blob/4eee5e088eb72ba00aa4b02a03a1bcf27b5d751b/src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala#L568-L575): the only related thing I could find in the compiler is the [`specializedClass `](https://github.com/scala/scala/blob/4eee5e088eb72ba00aa4b02a03a1bcf27b5d751b/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala#L80) map. The problem is that it only contain classes that are referenced by the program being compiled, the map is populated by method [`specializeClass `](https://github.com/scala/scala/blob/4eee5e088eb72ba00aa4b02a03a1bcf27b5d751b/src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala#L539). So if I want to use that functionality I'd have to invoke `specializeClass` for the `FunctionN` / `TupleN` types. Do you think that's the best way?

lrytz commented:
a temporary problem: using specialization's infrastructure wouldn't work for [`scala/runtime/java8/JFunction` types](https://github.com/scala/scala/blob/4eee5e088eb72ba00aa4b02a03a1bcf27b5d751b/src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala#L630-L631) (written in Java, manually specialized)

lrytz committed:
Harden methods to recognize method invocations to optimize

The previous methods to identify method invocations that can be
optimized, such as `isPredefAutoBox`, were String-based. Now we
obtain class and method signatures from symbols through the
BTypes infrastructure.

We also piggy-back on specialization's type transformer to create
all specialized subclasses of Tuple1/Tuple2. We'll do the same in
the future for FunctionN, but the current JFunctionN are written
in Java and specialized artisanally.

lrytz committed:
SI-8601 Don't treat newarray as dead code

Otherwise we lose the side effect of a `NegativeArraySizeException`.

A test for this case already exists (run/t8601b.scala), but it currently
enforces `-optimize -Ybackend:GenASM`, so it didn't trigger on the new
backend. However, PR #4814 was merged into 2.12.x and moved that test
over to the new backend and optimizer.  After merging the 2.12.x into
the current optimizer branch (push-pop elimination), the test started
failing.

Also disable the optimizer for `jvm/bytecode-test-example`: it counts
the number of null checks in a method, the optimizer (rightly) eliminates
one of the two.

lrytz committed:
Jump optimizations may enable more push-pop elimination

Jump optimization replaces an unnecessary conditional jump, e.g.
`IFNULL l; l: ...` by `POP`, which enables further push-pop elimination.

Also introduces a `-YoptTrace` flag that traces the progress of the
bytecode as it goes through local optimizations.

lrytz committed:
Merge remote-tracking branch 'upstream/2.12.x' into opt/elimBoxes

lrytz commented:
@retronym this PR should be ready now, the last 4 commits are new

retronym commented:
LGTM, awesome!

retronym commented:
I think the `-YoptTrace` option ought to accept an argument that lets one limit the output to a particular class / method pair. As it stands, it is only really useful after you've extracted the method of interest to a standalone test case. Feel free to do this in a followup PR, though.

retronym commented:
Here's a script to post-process the optimizer trace into a series of diffs (via a temp git repo). I've used your test case from 1a9e649 as the example.

https://gist.github.com/retronym/0efa5cea463aced52fae

retronym commented:
After looking at those diffs, I notice that `eliminateStaleStores` doesn't reduce `maxLocals`. This happens subsequently in `dce`, but I was wondering whether this is something that could/should be done more eagerly.

lrytz committed:
-YoptTrace takes a package/Class.method parameter

lrytz commented:
> the -YoptTrace option ought to accept an argument that lets one limit the output to a particular class / method pair

Added in 3dccf0e


https://github.com/lrytz/scala/commit/29d772d0655a3a5af79c3d8eca91bb33a82c7194#diff-70147974eed4d5f7faa546e42504ce64R94

> notice that eliminateStaleStores doesn't reduce maxLocals

This is intentional, see this comment: https://github.com/lrytz/scala/commit/29d772d0655a3a5af79c3d8eca91bb33a82c7194#diff-70147974eed4d5f7faa546e42504ce64R94

retronym commented:
LGTM
